adyenv6core/src/com/adyen/v6/facades/DefaultAdyenCheckoutFacade.java

```diff
@@ -1235,6 +1283,47 @@ private boolean isPosTimedOut(HttpServletRequest request) {
         }
     }
 
+    private OrderModel retrieveOrder(String orderCode) throws InvalidCartException {
+        OrderModel orderModel = getOrderRepository().getOrderModel(orderCode);
+        if (orderModel == null) {
+            //TODO change exception
+            throw new InvalidCartException("Order does not exist!");
+        }
+
+        getSessionService().removeAttribute(SESSION_LOCKED_CART);
+        getSessionService().removeAttribute(SESSION_PAYMENT_DATA);
+        getSessionService().removeAttribute(THREEDS2_FINGERPRINT_TOKEN);
+        getSessionService().removeAttribute(THREEDS2_CHALLENGE_TOKEN);
+        getSessionService().removeAttribute(PAYMENT_METHOD);
+
+        return orderModel;
+    }
+
+    private void restoreCartFromOrder(String orderCode) {
+        OrderModel orderModel = getOrderRepository().getOrderModel(orderCode);
+        if (orderModel == null) {
+            LOGGER.error("Could not restore cart to session, order not found!");
+            getCartService().removeSessionCart();
```

Comment:
which cart will be present in session during this execution is also important.

Reply:
Not sure if I understand your point, 
but I think I can remove the `getCartService().removeSessionCart();` and the current cart (if any) will remain in session... is that what you mean?

Reply:
Fixed that on branch PW-2530-restore-cart